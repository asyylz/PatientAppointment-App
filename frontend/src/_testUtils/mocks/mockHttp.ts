import useHttp from "../../hooks/useHttp/useHttp";

const deleteAppointmentMock = jest.fn();
// we have response conditon in Dashboard component to be able to empty the states
deleteAppointmentMock.mockResolvedValue({
  status: 204, // Simulate a successful deletion response with status 204
});

const updateAppointmentMock = jest.fn();
// we have response conditon in Dashboard component to be able to empty the states
updateAppointmentMock.mockResolvedValue({
  status: 'success',
});
const createAppointmentMock = jest.fn();
createAppointmentMock.mockResolvedValue({
  status: 'success',
});

const submitContactFormMock = jest.fn();
submitContactFormMock.mockResolvedValue({
  status: 'success',
});
const getDoctorWithAvailabilitiesMock = jest.fn();
getDoctorWithAvailabilitiesMock.mockResolvedValue({
  status: 'success',
});

(useHttp as jest.Mock).mockReturnValue({
  deleteAppointment: deleteAppointmentMock,
  updateAppointment: updateAppointmentMock,
  createAppointment: createAppointmentMock,
  getDoctorWithAvailabilities: getDoctorWithAvailabilitiesMock,
  submitContactForm: submitContactFormMock,
});

export {
  deleteAppointmentMock,
  updateAppointmentMock,
  createAppointmentMock,
  submitContactFormMock,
  getDoctorWithAvailabilitiesMock,
};
