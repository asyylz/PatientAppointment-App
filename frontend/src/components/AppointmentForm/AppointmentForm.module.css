.container {
    background: #F0F1F2;
    padding: 25px;
    border-radius: 20px;
    box-shadow: 0 10px 10px rgba(0, 0, 0, 0.15);
    margin: 1rem;
}


.title {
    font-weight: 600;
    margin: 20px 0 10px 0;
    position: relative;
}

.title::after {
    content: "";
    position: absolute;
    height: 4px;
    width: 100%;
    bottom: -10px;
    left: 0;
    border-radius: 5px;
    background: linear-gradient(0deg, rgba(34, 193, 195, 1) 23%, rgba(173, 217, 209, 1) 91%);
}


.container form {
    /* border: 1px solid blue; */
    display: flex;
    flex-direction: column;
    /* justify-content: center; */

}

.buttonContainer {
    display: flex;
    gap: 2rem;
    grid-column: span 2;
    margin-top: 2rem;
}

@media (max-width:768px) {
    .buttonContainer {
        flex-direction: column;
    }
}


@media (min-width:375px) and (max-width:834px) {
    .container {
        padding: 15px;
        /* border: 1px solid red; */

    }

    .buttonContainer {
        flex-direction: row;
        margin-top: 10px;
    }

    .buttonContainer button {
        width: 80px;
        font-size: 12px;
        height: 30px;
    }
}