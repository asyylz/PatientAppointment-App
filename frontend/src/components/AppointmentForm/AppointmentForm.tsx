import React, { useState } from 'react';
import classes from './AppointmentForm.module.css';
import useHttp from '../../hooks/useHttp/useHttp';
import CustomInput from '../CustomInput/CustomInput';
import ModalCustom from '../ModalCustom/ModalCustom';
import ReviewInput from '../ReviewInput/ReviewInput';
const reviewCriterias = ['Staff', 'Punctual', 'Helpful', 'Knowledge'];
import { dateForCustomInput } from './../../helper/generateDates';

interface AppointmentFormProps {
  setOpenModal: (openModal: string) => void;
  appointment: Appointment | SingleAppointmentForDoctor;
  isPatient?: boolean;
  userId?: string;
}

const AppointmentForm: React.FC<AppointmentFormProps> = ({
  setOpenModal,
  appointment,
  isPatient,
  userId,
}) => {
  const { updateAppointment, deleteAppointment, postReview } = useHttp();

  const [updatedAppointmentData, setUpdatedAppointmentData] =
    useState<AppointmentForUpdate>({
      appointmentDateAndTime: appointment.appointmentDateAndTime,
      reason: appointment.reason,
    });

  //console.log(typeof appointment._id);

  const [openModalDeleteOrComment, setOpenModalDeleteOrComment] =
    useState<string>('');

  const [appointmentDate, setAppointmentDate] = useState(
    appointment?.appointmentDateAndTime.split('T')[0]
  );
  const [appointmentTime, setAppointmentTime] = useState(
    appointment?.appointmentDateAndTime.split('T')[1]
  );
  const [ratingsAndComment, setRatingsAndComment] =
    useState<AttributesAndComment>({
      staff: 0,
      punctual: 0,
      helpful: 0,
      knowledge: 0,
      comments: '',
    });

  const handleChange = (
    e: React.ChangeEvent<
      HTMLInputElement | HTMLSelectElement | HTMLTextAreaElement
    >
  ) => {
    const { name, value } = e.target;

    if (name === 'appointmentDate') {
      setAppointmentDate(value);
      const formatted = `${value}T${appointmentTime}`;
      setUpdatedAppointmentData((prevValuesAppointment) => ({
        ...prevValuesAppointment,
        appointmentDateAndTime: formatted,
      }));
    } else if (name === 'appointmentTime') {
      setAppointmentTime(value);
      const formatted = `${appointmentDate}T${value}:00.000Z`;
      setUpdatedAppointmentData((prevValuesAppointment) => ({
        ...prevValuesAppointment,
        appointmentDateAndTime: formatted,
      }));
    } else if (name === 'comments') {
      setRatingsAndComment((prevValues) => ({
        ...prevValues,
        [name]: value,
      }));
    } else {
      setUpdatedAppointmentData((prevValuesAppointment) => ({
        ...prevValuesAppointment,
        [name]: value,
      }));
    }
  };

  const handleSubmit = async (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    const response = await updateAppointment(
      updatedAppointmentData,
      appointment?._id
    );
    if (response.status === 'success') {
      setOpenModal('');
    }
  };

  const handleDeleteAppointment = async (
    e: React.MouseEvent<HTMLButtonElement>
  ) => {
    e.preventDefault();
    const response = await deleteAppointment(appointment?._id);
    if (response?.status === 204) {
      setOpenModalDeleteOrComment('');
      setOpenModal('');
    }
  };

  const handlePostReview = async () => {
    const doctorId = isPatient
      ? (appointment as Appointment).doctorId._id
      : (appointment as SingleAppointmentForDoctor).doctorId;

    const response = await postReview({
      doctorId,
      userId,
      comments: ratingsAndComment.comments,
      attributes: {
        staff: ratingsAndComment.staff,
        punctual: ratingsAndComment.punctual,
        knowledge: ratingsAndComment.knowledge,
        helpful: ratingsAndComment.helpful,
      },
    });

    if (response.status === 'success') {
      setRatingsAndComment({
        staff: 0,
        punctual: 0,
        helpful: 0,
        knowledge: 0,
        comments: '',
      });
      setOpenModalDeleteOrComment('');
    }
  };
  const isPast = new Date(appointment.appointmentDateAndTime) < new Date();

  return (
    <>
      {/* /* -------------- Appointment Delete Notification Modal ------------- */}
      {openModalDeleteOrComment === 'open' && (
        <ModalCustom height="150px" width="300px">
          <p>
            {isPatient
              ? 'You are about to cancel your recent appointment?'
              : "Please confirm to delete the patient's appointment?"}
          </p>
          <div className="buttonContainer" style={{ flexDirection: 'row' }}>
            <button style={{ color: 'red' }} onClick={handleDeleteAppointment}>
              Confirm
            </button>
            <button
              style={{ color: 'blue' }}
              onClick={() => setOpenModalDeleteOrComment('')}
            >
              Cancel
            </button>
          </div>
        </ModalCustom>
      )}
      {/* /* -------------------- Comment Form -------------------- */}
      {openModalDeleteOrComment === 'comment' && (
        <ModalCustom height="auto" width="auto">
          <div className={`${classes.container} ${classes.comment}`}>
            <CustomInput
              readOnly
              value={`Dr. ${(appointment as Appointment).doctorId.firstName} ${
                (appointment as Appointment)?.doctorId.lastName
              }`}
            />
            <div className={classes.reviewWrapper}>
              {reviewCriterias.map((criteria: string) => {
                return (
                  <ReviewInput
                    attributeName={criteria}
                    ratingsAndComment={ratingsAndComment}
                    setRatingsAndComment={setRatingsAndComment}
                  />
                );
              })}
            </div>
            <textarea
              name="comments"
              rows={8}
              cols={36}
              onChange={handleChange}
            ></textarea>
            <div className={classes.buttonContainer}>
              <button onClick={handlePostReview}>Comment</button>
              <button
                onClick={() => {
                  setOpenModalDeleteOrComment('');
                  setRatingsAndComment({
                    staff: 0,
                    punctual: 0,
                    helpful: 0,
                    knowledge: 0,
                  });
                }}
              >
                Cancel
              </button>
            </div>
          </div>
        </ModalCustom>
      )}
      {/* /* ------------------ Appointment Form ------------------ */}
      <div className={classes.container}>
        <p className={classes.title}>Appointment Details</p>
        <form onSubmit={handleSubmit}>
          <CustomInput
            value={
              isPatient
                ? `Dr. ${(appointment as Appointment)?.doctorId?.firstName} ${
                    (appointment as Appointment)?.doctorId.lastName
                  }`
                : (appointment as SingleAppointmentForDoctor)?.patientId?.name
            }
            readOnly
          />
          <CustomInput
            placeHolder="Appointment Date"
            value={
              updatedAppointmentData?.appointmentDateAndTime &&
              dateForCustomInput(updatedAppointmentData?.appointmentDateAndTime)
            }
            type="date"
            name="appointmentDate"
            onChange={handleChange}
            readOnly={isPast}
          />

          <CustomInput
            placeHolder="Appointment Time"
            value={
              updatedAppointmentData?.appointmentDateAndTime &&
              updatedAppointmentData?.appointmentDateAndTime
                .split('T')[1]
                .slice(0, 5)
            }
            name="appointmentTime"
            type="time"
            step={1800}
            onChange={handleChange}
            readOnly={isPast}
          />
          <textarea
            className={classes.reason}
            placeholder="Please write your concerns..."
            value={updatedAppointmentData?.reason}
            name="reason"
            onChange={handleChange}
            rows={8}
            cols={36}
            style={isPast ? { opacity: 0.7, backgroundColor: 'lightgray' } : {}}
            readOnly={isPast}
          ></textarea>

          {!isPatient ? (
            <select
              name="status"
              id="status"
              onChange={handleChange}
              defaultValue={appointment?.status}
            >
              <option>Pelease choose a status</option>
              <option value="completed">Completed</option>
            </select>
          ) : (
            <CustomInput value={appointment?.status} readOnly />
          )}
          {!isPatient ? (
            <select
              name="referral"
              id="referral"
              onChange={handleChange}
              defaultValue={appointment?.referral.toString()}
            >
              <option>Pelease choose a referral status</option>
              <option value="true">Yes</option>
              <option value="false">No</option>
            </select>
          ) : (
            <CustomInput
              value={appointment?.referral ? 'Referral' : 'Not referral'}
              readOnly
            />
          )}

          <textarea
            className={classes.result}
            defaultValue={appointment?.diagnose}
            placeholder="Please write diagnoses..."
            name="diagnose"
            onChange={handleChange}
            rows={8}
            cols={36}
            readOnly={isPatient}
            style={
              isPatient ? { opacity: 0.7, backgroundColor: 'lightgray' } : {}
            }
          ></textarea>
          {appointment.status === 'completed' && isPast && isPatient && (
            <button
              onClick={(e) => {
                e.preventDefault();
                setOpenModalDeleteOrComment('comment');
              }}
              style={{
                fontSize: '12px',
                height: '30px',
                marginBottom: '5px',
              }}
            >
              Leave comment
            </button>
          )}

          <div className={classes.buttonContainer}>
            <button type="submit">Update</button>
            <button
              onClick={(e) => {
                e.preventDefault();
                setOpenModalDeleteOrComment('open');
              }}
            >
              Delete
            </button>
            <button onClick={() => setOpenModal('')}>Close</button>
          </div>
        </form>
      </div>
    </>
  );
};

export default AppointmentForm;
