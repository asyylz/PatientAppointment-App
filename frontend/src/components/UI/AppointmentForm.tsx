import React, { useState } from 'react';
import classes from './AppointmentForm.module.css';
import useHttp from '../../hooks/useHttp';
import CustomInput from './CustomInput';
import { toastErrorNotify } from '../../helper/ToastNotify';

interface AppointmentFormProps {
  setOpenModal: (openModal: string) => void;
  appointment: SingleAppointmentForDoctor | Appointment | undefined;
  isPatient: boolean;
}

const AppointmentForm: React.FC<AppointmentFormProps> = ({
  setOpenModal,
  appointment,
  isPatient,
}) => {
  const { updateAppointment, deleteAppointment } = useHttp();

  const [updatedAppointmentData, setUpdatedAppointmentData] = useState<
    object | undefined
  >();

  const [appointmentDate, setAppointmentDate] = useState(
    appointment?.appointmentDateAndTime.split('T')[0]
  );
  const [appointmentTime, setAppointmentTime] = useState(
    appointment?.appointmentDateAndTime.split('T')[1]
  );

  console.log('Doctorid', appointment?.doctorId);
  const handleChange = (
    e: React.ChangeEvent<
      HTMLInputElement | HTMLSelectElement | HTMLTextAreaElement
    >
  ) => {
    const { name, value } = e.target;

    if (name === 'appointmentDate') {
      setAppointmentDate(value);
      const formatted = new Date(`${value}T${appointmentTime}`);
      //console.log(formatted);
      setUpdatedAppointmentData((prevValuesAppointment) => ({
        ...prevValuesAppointment,
        appointmentDateAndTime: formatted,
      }));
    } else if (name === 'appointmentTime') {
      setAppointmentTime(value);
      const formatted = new Date(`${appointmentDate}T${value}:00.000Z`);
      setUpdatedAppointmentData((prevValuesAppointment) => ({
        ...prevValuesAppointment,
        appointmentDateAndTime: formatted,
      }));
    } else {
      setUpdatedAppointmentData((prevValuesAppointment) => ({
        ...prevValuesAppointment,
        [name]: value,
      }));
    }
  };

  const handleSubmit = async (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    const response = await updateAppointment(
      updatedAppointmentData,
      appointment?._id
    );
    if (response.status === 'success') {
      setOpenModal('');
    }
  };
  const handleDeleteAppointment = async (
    e: React.FormEvent<HTMLFormElement>
  ) => {
    e.preventDefault();
    const response = await deleteAppointment(appointment?._id);
    if (response?.status === 204) setOpenModal('');
    else toastErrorNotify('Appointment could not deleted');
  };

  return (
    <div className={classes.container}>
      <h1 className={classes.title}>Appointment Details</h1>
      <form onSubmit={handleSubmit}>
        <div className={classes.leftSection}>
          <CustomInput
            value={
              isPatient
                ? `Dr. ${appointment?.doctorId?.firstName} ${appointment?.doctorId.lastName}`
                : appointment?.patientId?.name
            }
            readOnly
          />
          <input
            placeholder="Appointment Date"
            // value={formatDateForUI(appointment?.appointmentDateAndTime)}
            defaultValue={appointment?.appointmentDateAndTime.split('T')[0]}
            type="date"
            name="appointmentDate"
            onChange={handleChange}
          />
          <input
            placeholder="Appointment Time"
            //value={appointment?.appointmentDateAndTime}
            defaultValue={appointment?.appointmentDateAndTime
              .split('T')[1]
              .slice(0, 5)}
            name="appointmentTime"
            type="time"
            step="1800"
            onChange={handleChange}
          />
          <textarea
            className={classes.reason}
            placeholder="Please write your concerns..."
            name="reason"
            defaultValue={appointment?.reason}
            onChange={handleChange}
            rows={8}
            cols={36}
          ></textarea>
        </div>

        <div className={classes.leftSection}>
          {!isPatient ? (
            <select
              name="status"
              id="status"
              onChange={handleChange}
              defaultValue={appointment?.status}
            >
              <option>Pelease choose a status</option>
              <option value="completed">Completed</option>
            </select>
          ) : (
            <CustomInput value={appointment?.status} readOnly />
          )}
          {!isPatient ? (
            <select
              name="referral"
              id="referral"
              onChange={handleChange}
              defaultValue={appointment?.referral.toString()}
            >
              <option>Pelease choose a referral status</option>
              <option value="true">Yes</option>
              <option value="false">No</option>
            </select>
          ) : (
            <CustomInput
              value={appointment?.referral ? 'Referral' : 'Not referral'}
              readOnly
            />
          )}

          <textarea
            className={classes.result}
            defaultValue={appointment?.diagnose}
            placeholder="Please write diagnoses..."
            name="diagnose"
            onChange={handleChange}
            rows={8}
            cols={36}
            readOnly={isPatient}
            style={
              isPatient ? { opacity: 0.7, backgroundColor: 'lightgray' } : {}
            }
          ></textarea>
        </div>
        <div className={classes.buttonContainer}>
          {' '}
          <button type="submit">Update</button>
          <button onClick={handleDeleteAppointment}>Cancel</button>
          <button onClick={() => setOpenModal('')}>Close</button>
        </div>
      </form>
    </div>
  );
};

export default AppointmentForm;
