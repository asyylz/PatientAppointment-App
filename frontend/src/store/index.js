import { configureStore } from '@reduxjs/toolkit';

const store = configureStore({
  reducer: { departments: departmentsSlice.reducer }, // root reducer
});

export default store;
