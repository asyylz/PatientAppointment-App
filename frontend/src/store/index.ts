import { configureStore, combineReducers } from '@reduxjs/toolkit';
import {
  persistReducer,
  persistStore,
  FLUSH,
  REHYDRATE,
  PAUSE,
  PERSIST,
  PURGE,
  REGISTER,
} from 'redux-persist';
import departmentsReducer from './departments-slice';
import doctorsReducer from './doctors-slice';
import reviewsReducer from './reviews-slice';
import currentUserReducer from './currentUser-slice';
//import { useDispatch } from 'react-redux';
import searchReducer from './search-slice';
import appointmentsForDoctorReducer from './appointmentsForDoctor-slice';
import appointmentsForPatientReducer from './appointmentsForPatient-slice';
import storage from 'redux-persist/lib/storage'; // defaults to localStorage for web

// const rootPersistConfig = {
//   key: 'root',
//   storage,
// };

// const persistedReducer = persistReducer(rootPersistConfig, currentUserReducer);

// const store = configureStore({
//   reducer: {
//     currentUser: persistedReducer,
//     departments: departmentsReducer,
//     doctors: doctorsReducer,
//     reviews: reviewsReducer,
//     search: searchReducer,
//     appointmentsForDoctor: appointmentsForDoctorReducer,
//     appointmentsForPatient: appointmentsForPatientReducer,
//   },
//   middleware: (getDefaultMiddleware) =>
//     getDefaultMiddleware({
//       serializableCheck: {
//         ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
//         ignoredActionPaths: ['meta.arg', 'payload.timestamp'],
//         ignoredPaths: ['items.dates'],
//       },
//     }),
//   devTools: process.env.NODE_ENV !== 'production',
// });

// const rootPersistConfig = {
//   key: 'root',
//   storage,
//   whitelist: ['currentUser', 'doctors'], // Only persist the 'currentUserSlice' slice
// };

const authPersistConfig = {
  key: 'currentUser',
  storage: storage,
  // blacklist: ['status', 'error'],
};
const doctorPersistConfig = {
  key: 'doctors',
  storage: storage,
  // blacklist: ['status','error']
};
export const allReducers = combineReducers({
  currentUser: persistReducer(authPersistConfig, currentUserReducer),
  departments: departmentsReducer,
  doctors: persistReducer(doctorPersistConfig, doctorsReducer),
  reviews: reviewsReducer,
  search: searchReducer,
  appointmentsForDoctor: appointmentsForDoctorReducer,
  appointmentsForPatient: appointmentsForPatientReducer,
});

//const persistedReducer = persistReducer(rootPersistConfig, allReducers);

const store = configureStore({
  reducer: allReducers,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      serializableCheck: {
        ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
      },
    }),
  devTools: process.env.NODE_ENV !== 'production',
});

export type AppDispatch = typeof store.dispatch;
//export const useAppDispatch = useDispatch.withTypes<AppDispatch>();
export type AppStore = typeof store;
export type RootState = ReturnType<AppStore['getState']>;

export const persistor = persistStore(store);

export default store;
