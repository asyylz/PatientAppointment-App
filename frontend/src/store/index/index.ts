import { configureStore, combineReducers } from '@reduxjs/toolkit';
import {
  persistReducer,
  persistStore,
  FLUSH,
  REHYDRATE,
  PAUSE,
  PERSIST,
  PURGE,
  REGISTER,
} from 'redux-persist';
import departmentsReducer from '../departments-slice/departments-slice';
import doctorsReducer from '../doctors-slice/doctors-slice';
import reviewsReducer from './../reviews-slice/reviews-slice';
import currentUserReducer from '../currentUser-slice/currentUser-slice';
import searchReducer from '../search-slice/search-slice';
import appointmentsForDoctorReducer from '../appointmentsForDoctor-slice/appointmentsForDoctor-slice';
import appointmentsForPatientReducer from '../appointmentsForPatient-slice/appointmentsForPatient-slice';
import storage from 'redux-persist/lib/storage'; // defaults to localStorage for web

const authPersistConfig = {
  key: 'currentUser',
  storage: storage,
  // blacklist: ['status', 'error'],
};

const doctorPersistConfig = {
  key: 'doctors',
  storage: storage,
  // blacklist: ['status','error']
};

export const allReducers = combineReducers({
  currentUser: persistReducer(authPersistConfig, currentUserReducer),
  departments: departmentsReducer,
  doctors: persistReducer(doctorPersistConfig, doctorsReducer),
  reviews: reviewsReducer,
  search: searchReducer,
  appointmentsForDoctor: appointmentsForDoctorReducer,
  appointmentsForPatient: appointmentsForPatientReducer,
});

const store = configureStore({
  reducer: allReducers,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      serializableCheck: {
        ignoredActions: [
          FLUSH,
          REHYDRATE,
          PAUSE,
          PERSIST,
          PURGE,
          REGISTER,
          'appointmentsForPatient/fetchWithId/rejected',
          'appointmentsForDoctor/fetchWithId/rejected',
          'departments/fetch/rejected',
          'doctors/fetch/rejected',
        ],
      },
    }),
  devTools: process.env.NODE_ENV !== 'production',
});

export function setupStore(reducer?: any, preloadedState?: Partial<RootState>) {
  //console.log('from index store', preloadedState); // i can see my preloaded state here  with mock data
  const store = configureStore({
    reducer: reducer,
    preloadedState,
  });

  //console.log(store.getState()); // but still cant see current user but can see the rest of the states
  return store;
}

export type AppDispatch = typeof store.dispatch;
export type AppStore = typeof store;
export type RootState = ReturnType<AppStore['getState']>;

export const persistor = persistStore(store);

export default store;
