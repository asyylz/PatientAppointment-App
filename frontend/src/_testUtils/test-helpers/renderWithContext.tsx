import React, { PropsWithChildren } from 'react';
import { render } from '@testing-library/react';
import type { RenderOptions } from '@testing-library/react';
import { Provider } from 'react-redux';
import { setupStore } from '../../store/index';
import type { AppStore, RootState } from '../../store/index';

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  reducer?: any;
  preloadedState?: Partial<RootState>;
  store?: AppStore;
}

export function renderWithProviders(
  ui: React.ReactElement,
  extendedRenderOptions: ExtendedRenderOptions = {}
) {
  const { preloadedState, reducer } = extendedRenderOptions;
  // Ensure that the store is being correctly initialized with preloadedState
  const store = setupStore(reducer, preloadedState);
  //console.log(store.getState()); // Verify the state of the store

  const Wrapper = ({ children }: PropsWithChildren) => (
    <Provider store={store}>{children}</Provider>
  );

  // Return an object with the store and all of RTL's query functions
  return {
    store,
    ...render(ui, { wrapper: Wrapper, ...extendedRenderOptions }),
  };
}
