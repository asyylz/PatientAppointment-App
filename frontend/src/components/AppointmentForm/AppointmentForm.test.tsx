import { combineReducers } from 'redux';
import AppointmentForm from './AppointmentForm';
import currentUserReducer from './../../store/currentUser-slice/currentUser-slice.ts';
import reviewsReducer from './../../store/reviews-slice/reviews-slice.ts';
import { AppStore } from '../../store/index';
import { renderWithProviders } from '../../_testUtils/test-helpers/renderWithContext';
import { PersistPartial } from 'redux-persist/es/persistReducer';
import { act, fireEvent, screen } from '@testing-library/react';
import { postReviewMock } from '../../_testUtils/mocks/mockHttp.ts';

jest.mock('../../hooks/useHttp/useHttp');
jest.mock('../../services/axiosInterceptors');

const initialCurrentUserState: CurrentUser = {
  status: 'login success',
  token: 'someToken',
  userData: {
    _id: '6673662fbd42a966b75dec92',
    name: 'Alice',
    email: 'alice@test.com',
    role: 'patient',
    __v: 0,
    DOB: '1986-01-22T00:00:00.000Z',
    image:
      'https://patient-appointment-system.s3.eu-west-2.amazonaws.com/1722605356602-asiye%20serife.jpeg',
    policy: true,
  },
  error: null,
};

const appointment = {
  _id: '66bc6c89a0c6ea472a2fcccf',
  doctorId: {
    _id: '665f0f2959c971659af920ca',
    firstName: 'Bowen',
    lastName: 'Chan',
    departmentId: '6660fbd12e3c00fe18cfceef',
  },
  patientId: '6673662fbd42a966b75dec92',
  appointmentDateAndTime: '2024-08-14T15:30:00.000Z',
  reason: 'test1',
  diagnose: null,
  status: 'completed',
  referral: false,
  _v: 0,
};

const reducers = combineReducers({
  currentUser: currentUserReducer,
  reviews: reviewsReducer,
  // other reducers...
});

describe('AppointmentForm Component', () => {
  let store: AppStore;
  let asFragment: () => DocumentFragment;

  const mockSetOpenModal = jest.fn().mockImplementation(() => '');

  const initialProps = {
    appointment: appointment as unknown as Appointment,
    userId: 'mockUserId',
    setOpenModal: mockSetOpenModal,
    isPatient: true,
  };

  beforeEach(() => {
    mockSetOpenModal.mockClear();
    const result = renderWithProviders(
      <>
        <div id="modal" data-testid="modal"></div>
        <AppointmentForm {...initialProps} />
      </>,
      {
        reducer: reducers,
        preloadedState: { currentUser: initialCurrentUserState } as Partial<{
          currentUser: CurrentUser & PersistPartial;
        }>,
      }
    );
    store = result.store;
    asFragment = result.asFragment;
    jest.spyOn(store, 'dispatch');
  });

  afterEach(() => {
    jest.clearAllMocks();
  });
  /* -------------------------- - ------------------------- */
  it('1--Matches the DOM snapshot', async () => {
    expect(asFragment()).toMatchSnapshot();
  });
  /* -------------------------- - ------------------------- */
  it('2--Input changes should be reflected in DOM', () => {
    const time = screen.getByPlaceholderText('Appointment Time');
    const date = screen.getByPlaceholderText('Appointment Date');
    const reason = screen.getByPlaceholderText('Please write your concerns...');
    fireEvent.change(time, { target: { value: '11:30' } });
    fireEvent.change(date, { target: { value: '2024-08-20' } });
    fireEvent.change(reason, { target: { value: 'new updated reason' } });

    expect(screen.queryByText('11:30'));
    expect(screen.queryByText('2024-08-20'));
    expect(screen.queryByText('new updated reason'));
  });
  /* -------------------------- - ------------------------- */
  it('3--PostReview should be called with corret arguments', async () => {
    const leaveCommentButton = screen.getByRole('button', {
      name: 'Leave comment',
    });
    fireEvent.click(leaveCommentButton);

    const comments = screen.getByPlaceholderText('Leave your comments here...');
    fireEvent.change(comments, { target: { value: 'new comments' } });
    expect(screen.queryByText('new comments'));

    const commentButton = screen.getByRole('button', {
      name: 'Comment',
    });
    fireEvent.click(commentButton);

    // Wait for deleteAppointment to be called
    await act(async () => {
      expect(postReviewMock).toHaveBeenCalledWith({
        attributes: { helpful: 0, knowledge: 0, punctual: 0, staff: 0 }, // these values come from initial props
        comments: 'new comments',
        doctorId: '665f0f2959c971659af920ca',
        userId: 'mockUserId',
      });
      expect(postReviewMock).toHaveBeenCalledTimes(1);
    });
  });
  /* -------------------------- - ------------------------- */
  it('4--After cancelling review   related states should be emptied', () => {
    const leaveCommentButton = screen.getByRole('button', {
        name: 'Leave comment',
      });
      fireEvent.click(leaveCommentButton);
    const cancelButton = screen.getByRole('button', { name: 'Cancel' });
    fireEvent.click(cancelButton);
    expect(screen.queryByText('Leave your comments here...')).not.toBeInTheDocument();
  });
  it('5--Closing AppointmentForm', () => {
    const closeButton = screen.getByRole('button', { name: 'Close' });
    fireEvent.click(closeButton);
    expect(screen.queryByText('Please write diognoses...')).not.toBeInTheDocument();
  });
});
