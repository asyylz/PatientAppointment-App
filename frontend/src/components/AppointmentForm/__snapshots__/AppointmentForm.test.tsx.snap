// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppointmentForm Component 1--Matches the DOM snapshot 1`] = `
<DocumentFragment>
  <div
    data-testid="modal"
    id="modal"
  />
  <div
    class="container"
  >
    <p
      class="title"
    >
      Appointment Details
    </p>
    <form>
      <div
        class="wrapper input"
        style="opacity: 0.7;"
      >
        <input
          readonly=""
          style="background-color: lightgray;"
          value="Dr. Bowen Chan"
        />
      </div>
      <div
        class="wrapper input"
        style="opacity: 0.7;"
      >
        <input
          name="appointmentDate"
          placeholder="Appointment Date"
          readonly=""
          style="background-color: lightgray;"
          type="date"
          value="2024-08-14"
        />
      </div>
      <div
        class="wrapper input"
        style="opacity: 0.7;"
      >
        <input
          name="appointmentTime"
          placeholder="Appointment Time"
          readonly=""
          step="1800"
          style="background-color: lightgray;"
          type="time"
          value="15:30"
        />
      </div>
      <textarea
        class="reason"
        cols="36"
        name="reason"
        placeholder="Please write your concerns..."
        readonly=""
        rows="8"
        style="opacity: 0.7; background-color: lightgray;"
      >
        test1
      </textarea>
      <div
        class="wrapper input"
        style="opacity: 0.7;"
      >
        <input
          readonly=""
          style="background-color: lightgray;"
          value="completed"
        />
      </div>
      <div
        class="wrapper input"
        style="opacity: 0.7;"
      >
        <input
          readonly=""
          style="background-color: lightgray;"
          value="Not referral"
        />
      </div>
      <textarea
        class="result"
        cols="36"
        name="diagnose"
        placeholder="Please write diagnoses..."
        readonly=""
        rows="8"
        style="opacity: 0.7; background-color: lightgray;"
      />
      <button
        style="font-size: 12px; height: 30px; margin-bottom: 5px;"
      >
        Leave comment
      </button>
      <div
        class="buttonContainer"
      >
        <button
          type="submit"
        >
          Update
        </button>
        <button>
          Delete
        </button>
        <button>
          Close
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
